helper/metricsutil/gauge_process.go

```diff
@@ -0,0 +1,236 @@
+func (m *ClusterMetricSink) NewGaugeCollectionProcess(
+	key []string,
+	id []Label,
+	collector GaugeCollector,
+	logger log.Logger,
+) (*GaugeCollectionProcess, error) {
+	return m.newGcpWithClock(
+		key,
+		id,
+		collector,
+		logger,
+		defaultClock{},
+	)
+}
+
+// test version allows an alternative clock implementation
+func (m *ClusterMetricSink) newGcpWithClock(
+	key []string,
+	id []Label,
+	collector GaugeCollector,
+	logger log.Logger,
+	clock clock,
+) (*GaugeCollectionProcess, error) {
+	process := &GaugeCollectionProcess{
+		Stop:             make(chan struct{}, 1),
+		stopped:          make(chan struct{}, 1),
+		key:              key,
+		labels:           id,
+		collector:        collector,
+		sink:             m,
+		originalInterval: m.GaugeInterval,
+		currentInterval:  m.GaugeInterval,
+		logger:           logger,
+		clock:            clock,
+	}
+	return process, nil
+}
+
+// delayStart randomly delays by up to one extra interval
+// so that collection processes do not all run at the time time.
+// If we knew all the procsses in advance, we could just schedule them
+// evenly, but a new one could be added per secret engine.
+func (p *GaugeCollectionProcess) delayStart() bool {
+	randomDelay := time.Duration(rand.Intn(int(p.currentInterval)))
+	// FIXME: a Timer might be better, but then we'd have to simulate
+	// one of those too?
+	delayTick := p.clock.NewTicker(randomDelay)
+	defer delayTick.Stop()
+
+	select {
+	case <-p.Stop:
+		return true
+	case <-delayTick.C:
+		break
+	}
+	return false
+}
+
+// resetTicker stops the old ticker and starts a new one at the current
+// interval setting.
+func (p *GaugeCollectionProcess) resetTicker() {
+	if p.ticker != nil {
+		p.ticker.Stop()
+	}
+	p.ticker = p.clock.NewTicker(p.currentInterval)
+}
+
+// collectAndFilterGauges executes the callback function,
+// limits the cardinality, and streams the results to the metrics sink.
+func (p *GaugeCollectionProcess) collectAndFilterGauges() {
+	// Run for only an allotted amount of time.
+	timeout := time.Duration(collectionBound * float64(p.currentInterval))
+	ctx, cancel := context.WithTimeout(context.Background(),
+		timeout)
+	defer cancel()
+
+	start := p.clock.Now()
+	values, err := p.collector(ctx)
+	end := p.clock.Now()
+	duration := end.Sub(start)
+
+	// Report how long it took to perform the operation.
+	p.sink.AddDurationWithLabels([]string{"metrics", "collection"},
+		duration,
+		p.labels)
+
+	// If over threshold, back off by doubling the measurement interval.
+	// Currently a restart is the only way to bring it back down.
+	threshold := time.Duration(collectionTarget * float64(p.currentInterval))
+	if duration > threshold {
+		p.logger.Warn("gauge collection time exceeded target", "target", threshold, "actual", duration, "id", p.labels)
+		p.currentInterval *= 2
+		p.resetTicker()
+	}
+
+	if err != nil {
+		p.logger.Error("error collecting gauge", "id", p.labels, "error", err)
```

Comment:
I worry a little about this log message being too spammy. Do we want to use lower severity? Throttle? Replace with an error metric?

Personally regardless of whether we change the logging, I like having error metrics for my metrics collections. That way when creating a dashboard or alert rule, I can easily take into account if the metrics are dubious due to a collection error.

Reply:
I added an error count.

I think the frequency of error messages will not be too high to worry about, in this case, but it might become a concern if there is a storage outage. I suspect there will be many more error messages in that case anyway?